### PR TITLE
Utilise ES6 features in controllers.

### PR DESCRIPTION
**Pull request by joefitter**
- base-controller, date-controller and error controller are now ES6 classes
- Arrow functions added where applic
- Super method calls simplified by using `super` keyword
- Block scoping with `const` and `let`

### CODE REVIEW

**Comment by joefitter** [comment]
@easternbloc @JoeChapman @gxxm mind having a look over when you get a sec?

**Comment by easternbloc** [comment]
Looks good but can we hold off merging this till we clear the other PR's.

**Comment by JoeChapman** [comment]
@joefitter needs a rebase

**Comment by joefitter** [comment]
@JoeChapman @easternbloc @gxxm @daniel-ac-martin I have rebased with latest changes - are you able to have a 😍 ?

**Comment by JoeChapman** [comment]
@joefitter something @easternbloc said about debugging. All arrow functions are anonymous so stack traces will be a PITA to read. debug could help us to alleviate this issue, thought?
Would it be fairly quick to add debug and some debug statements into this PR, PR debug separately?

**Comment by joefitter** [comment]
I have mixed feelings about littering our codebase with debug statements, and also what exactly would we be debugging? Just adding a `debug('functionName')` inside anonymous functions? Arrow functions are anonymous and are part of the language, I guess it's a choice to either use them or name every function. I do see the point of adding debug statements to complex parts of logic, but just think if we want to be naming every function we shouldn't be using arrow functions at all

**Comment by easternbloc** [comment]
> I have mixed feelings about littering our codebase with debug statements, and also what exactly would we be debugging? Just adding a debug('functionName') inside anonymous functions?

Yeah tend to agree. I don't know what the solution is tbh I just think we're regressing somewhat going from named functions (helpful for working out a stack trace) to fully anon (can be a real PITA).

I still like my named functions they save you a lot of time when you're knee deep in the anon hell.

> but just think if we want to be naming every function we shouldn't be using arrow functions at all

Probably that.

**Comment by easternbloc** [comment]
I do want debug where debug is useful in our modules though. But that's a different issue.

**Comment by joefitter** [comment]
@easternbloc I know what you mean, but it seems a shame to not use one of the best (IMO) parts of ES6 just because they aren't named in stack traces. Almost makes me think why are we using ES6 at all - as pretty much the only difference will be the use of `const` and `let` in place of `var`. 

Personally I don't have any issue with anonymous/arrow functions being used as callbacks provided they are used sensibly and not excessively deep so you can still see the named caller method/function in stack traces.

In the case of iteratee functions passed to lodash methods you can see the name of the lodash method in the stack trace anyway - I don't see the point of doing `_.map(items, function mapItems(items) { ... } )` as to me it feels like we are going against the spirit of the language by being overly verbose and cutting out anon functions entirely.

I do agree with you in cases where we are doing `const func = () => { ... }` as these should really still use a `function` declaration.

@JoeChapman @easternbloc @daniel-ac-martin @gxxm discuss! :)

**Comment by JoeChapman** [comment]
@joefitter @easternbloc @daniel-ac-martin @gavboulton I don't think anyone is suggesting we use `debug('function_name')`. But a debug with useful info where necessary would help identify the area in which the code was failing. With appropriate debug statements (not simply printing function names) we wouldn't need to rely on a stack trace at all.

I want to continue using arrow functions. I want to use debug too. We should not use debug statements to print the name of a function.

**Comment by joefitter** [comment]
Yeah I didn't think you meant that, +1 for arrow functions, +1 for debug

**Comment by joefitter** [comment]
I think debug should be a separate issue/PR though.

**Comment by JoeChapman** [comment]
I agree with you @joefitter I shouldn't have brought this up. Let's merge this PR.

**Comment by JoeChapman** [comment]
@easternbloc do we still need to wait?

**Comment by easternbloc** [comment]
I'm still not sure what I think about => vs traceability when things go wrong. Call me old fashioned but I like to be able to read a stack trace. That being said:

> I don't see the point of doing _.map(items, function mapItems(items) { ... } )

Is OTT so maybe we just try and trust ourselves not to do dumb things with arrows.

**Comment by easternbloc** [comment]
Maybe have it at the forefront of your mind when reviewing PR's though. I dunno. Whatevs I'll go back in my box now.

**Comment by JoeChapman** [comment]
I think we should get debug in, in another PR and then forefront of our minds when PR'ing, reviewing should be are we adequately and appropriately debugged using `debug` when running dev, ok?

**Comment by joefitter** [comment]
That being said we happy to merge this? @easternbloc @JoeChapman

**Comment by easternbloc** [comment]
@joefitter only if you promise me it'll make your weekend?

**Comment by JoeChapman** [comment]
lol. LGTM, @easternbloc? If you are happy to merge this, merge this.

**Comment by easternbloc** [comment]
Let me just cast my 👀 over it again...

**Comment by joefitter** [comment]
promise

File: .jscsrc.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "preset": "google",
+  "esnext": true,
```
Comment (JoeChapman) [inline .jscsrc.json:3]: update jscs and this is included by default (the options has been removed in 3.0)

File: lib/base-controller.js [hunk 2]
```diff
@@ -1,103 +1,98 @@
...
+    super(options);
+    this.confirmStep = options.confirmStep || '/confirm';
+  }
 
-function getErrorLength() {
-  var errors = this.getErrors.apply(this, arguments);
-  var errorLength = Object.keys(errors).length;
+  getNextStep(req, res) {
+    let next = super.getNextStep(req, res);
+    const forks = this.options.forks || [];
```
Comment (easternbloc) [inline lib/base-controller.js:15]: in my head lets go below consts. My head is weird.

Comment (joefitter) [inline lib/base-controller.js:15]: I think I agree, I was just find and replacing - lazy

File: .jscsrc.json (hunk 1 above)
Comment (joefitter) [inline .jscsrc.json:3]: would updating JSCS be a separate PR?

Comment (easternbloc) [inline .jscsrc.json:3]: It will now :)
